### ISSUE TITLE
Initialize semaphore permits programattically

### ISSUE THREAD

**Issue by manoojm** [hazelcast/hazelcast-nodejs-client#509]
I have a requirement to create a semaphore with just exactly one permit. I have multiple clients trying to create the semaphore and a reliable way to initialize the semaphore exactly once during the create time with 1 permit. Any suggestions?

**Comment by puzpuzpuz** [comment]
Why don't you want to use lock (`ILock`) for this purpose? It should behave like semaphore with one permit.

An important note. In IMDG 3.12 older implementations of lock and semaphore were deprecated and new linearizable data structures were introduced as a part of the new [CP Subsystem](https://docs.hazelcast.org/docs/latest/manual/html-single/index.html#cp-subsystem). They're safer than older data structures as they're based on Raft consensus protocol. However, Node.js client doesn't support them yet.

**Comment by manoojm** [comment]
We have moved to latest client and hazelcast member in our deployment and we don't notice any issues. Your statement on "nodejs doesn't support them yet" worries me.

Coming to ILock, in nodejs, it's single threaded, so the same thread can get the lock and perform the same task multiple times. I need to prevent that as well which is why we decided to use Semaphore but faced the difficulty.

**Comment by puzpuzpuz** [comment]
> We have moved to latest client and hazelcast member in our deployment and we don't notice any issues.

Old lock and semaphore are deprecated, but not removed yet. That's why your code keeps working with the latest IMDG version.

> Coming to ILock, in nodejs, it's single threaded, so the same thread can get the lock and perform the same task multiple times. I need to prevent that as well which is why we decided to use Semaphore but faced the difficulty.

Just as `ISemaphore`, `ILock` is a data structure that is stored on server-side, so in the Node.js client case it has nothing to do with multi-threaded programming. Lock data structure guarantees that only a single lock can be acquired among all of your Node.js processes, not within a single Node.js process. You just need to guard your "critical section" with `lock.lock()` call, like the following:

```javascript
const lock = hz.getLock('my-distributed-lock');
try {
  await lock.lock();
  // your guarded code goes here
} finally {
  await lock.unlock();
}
```

**Comment by manoojm** [comment]
I dont think u understood my requirement. I want to not run the same piece of code even in the same nodejs process. Lock helps me not run the same piece of code concurrently from multiple nodejs processes. But, the same nodejs process can keep running the same piece of code again and again because it has the lock. This is where semaphore can help me.

**Comment by puzpuzpuz** [comment]
Let me rephrase. An `ISemaphore` with a single permit is equal to an `ILock`: it guarantees that a single lock can be acquired among all of your Node.js processes (and, of course, within a single Node.js process). It's also written in the 3.11.2 reference manual:

> ISemaphore with a single permit may be considered as a lock. Unlike the locks, when semaphores are used, any thread can release the permit, and semaphores can have multiple permits.

(note: the part that mentions threads is not applicable to Node.js client)

But if you still want to use `ISemaphore`s, you can configure them (including, permits) with XML configuration on server side. In this case you won't need to have `.init()` call in your Node.js code. See [this section](https://docs.hazelcast.org/docs/3.11.2/manual/html-single/index.html#configuring-semaphore) of the reference manual.

**Comment by puzpuzpuz** [comment]
@manoojm does my recent message answer your question? If yes, I'll close the issue.

**Comment by manoojm** [comment]
Yes, it does. I have closed the issue. Thanks.
